common/src/main/java/org/apache/comet/vector/CometVector.java

```diff
@@ -79,14 +79,23 @@ public Decimal getDecimal(int i, int precision, int scale) {
     } else {
       byte[] bytes = getBinaryDecimal(i);
       BigInteger bigInteger = new BigInteger(bytes);
-      BigDecimal javaDecimal = new BigDecimal(bigInteger, scale);
```

Comment:
For decimals with small precisions, the `BigDecimal` is unnecessary and wasting time/memory.

Reply:
This is only for precision  > 18. `Decimal` internally keeps the value either as BigDecimal or long.  By assigning to BigInteger the Decimal implementation will check if the value will fit in a long and if not then convert to a BigDecimal. But we have already made the check to see if the value will fit into a long so we can save that check by converting to BigDecimal directly.

Reply:
If use decimal128 is true, we always go this no matter precision. That's said we always use BigDecimal even for int/long cases.

Reply:
And for native execution, we always enable use_decimal128.

Reply:
True. 
Looks like Decimal.apply(BigInteger) has a fairly expensive check to see how many bits are in use and see of the value will fit into an integer/long but it might be worthy it downstream. 
Are you able to see a difference in the micro benchmark with useDecimal128 enabled?

Reply:
I have not run the microbenchmark yet but just looked at the Decimal.apply and thought this might be helpful. Currently I also wonder why it fails (overflow) on converting an integer 12345 to Decimal of precision 5 and scale 2.

Reply:
Ok. Let's see if this might improve things. Provided ci succeeds.